datafusion-federation/src/optimize.rs

```diff
@@ -1,27 +1,59 @@
 use datafusion::{
-    common::tree_node::{Transformed, TreeNode, TreeNodeRewriter},
+    common::tree_node::{Transformed, TransformedResult, TreeNode, TreeNodeRewriter},
     error::Result,
-    logical_expr:: LogicalPlan,
+    execution::{SessionState, SessionStateBuilder},
+    logical_expr::LogicalPlan,
     optimizer::{
-        optimizer::ApplyOrder, push_down_filter, OptimizerConfig, OptimizerContext, OptimizerRule
-    }
+        optimize_projections::OptimizeProjections, optimizer::ApplyOrder,
+        push_down_filter::PushDownFilter, OptimizerConfig, OptimizerRule,
+    },
+    prelude::SessionConfig,
 };
 
-pub(crate) fn optimize_plan(plan: LogicalPlan) -> Result<LogicalPlan> {
-    let push_down_filter_rule = push_down_filter::PushDownFilter::new();
-    // `push_down_filter` does not use config so it can be default
-    let optimizer_config = OptimizerContext::default();
+pub(crate) struct Optimizer {
+    state: SessionState,
+    push_down_filter: PushDownFilter,
+    optimize_projections: OptimizeProjections,
+}
 
-    let res = match push_down_filter_rule.apply_order() {
-        Some(apply_order) => plan.rewrite(&mut Rewriter::new(
-            apply_order,
-            &push_down_filter_rule,
-            &optimizer_config,
-        )),
-        None => optimize_plan_node(plan, &push_down_filter_rule, &optimizer_config),
-    };
+impl Default for Optimizer {
+    fn default() -> Self {
+        // `push_down_filter` and `optimize_projections` does not use config (except `optimize_projections_preserve_existing_projections`) so it can be default
+        let config =
+            SessionConfig::new().with_optimize_projections_preserve_existing_projections(true);
+        // `SessionState` implements `OptimizerConfig` and allows specifying required configuration for optimization rules
```

Comment:
Empty SessionState is cheap to create and use. Creating custom struct that implements OptimizerConfig will require adding `chrono` dependency